### PR TITLE
[Transformers backend] Find attention with a fuser and attach vLLM's layer to it

### PR DESCRIPTION
**Pull request by bohnstingl**
## Purpose

Teach the Transformers modeling backend to find each layer's attention with a fuser, attach
vLLM's attention layer to the HF module that dispatches it, and resolve the softmax scale at
construction. `create_attention_instances` built one `Attention` per layer into
`self.attention_instances`, a plain `dict`. `nn.Module.__setattr__` does not register a dict, so
none of those layers was part of the module tree, and `vllm_attention_forward` reached its own
with `attention_instances[module.layer_idx]`. The MLA path already worked around half of this by
attaching its instance as `mla_module._vllm_mla_attn`, precisely so the layer appeared in
`named_modules()` and ran its `process_weights_after_loading`.

**`AttentionFuser` (new).** A module dispatches attention if its `forward` binds
`ALL_ATTENTION_FUNCTIONS.get_interface(...)` to a local and calls it exactly once. That is read
from the forward's source, not from the `fx` trace: `trace` is deliberately partial-tolerant and
returns whatever graph it managed, so a missing interface node does not mean the module has no
dispatch. Keying off `self.fusers` would be wrong for the same reason a QKV fuser is not the
right handle -- a model can dispatch through the interface without its projections fusing at all.
The fuser also carries the `scaling=` expression it found, and answers `layer_index` from the
module's own `layer_idx`.

**Fusers are now plural per module.** `get_fusers` returns every fuser that applies: at most one
that redefines the forward (`redefines_forward`, mutually exclusive because each rewrite starts
from the original source) plus any number that leave it alone. `AttentionFuser` is the first of
the latter kind, so a Llama attention comes back as `[QKVFuser, AttentionFuser]` and a Gemma 4
attention -- whose QKV fusion is rejected -- as `[AttentionFuser]`. This also drops a special case:
an `MLAFuser` that fails `validate` used to be rebuilt as a bare attention marker, and now simply
falls out of the list while the `AttentionFuser` stands on its own.

This fixes two latent bugs and removes one requirement:

1. **`impl.scale` was written per forward, which one compiled artifact per layer class cannot
   preserve.** Every layer was built with the Llama default `head_size**-0.5` and corrected from
   the HF module's `scaling` kwarg on every call. That correction only survives because
   `module.layer_idx` is an `int` baked into the graph, so Dynamo compiles the stack layer by
   layer. The moment one artifact is reused across layers -- the point of `_USE_LAYERNAME` /
   `LayerName` hoisting -- only the first layer's Python frame runs and layers 1..N silently keep
   the default. Wrong for any model whose scale is not derived from `head_size`. The scale is now
   read from the `scaling=` argument the module hands the interface, once, at construction.
   Reading the argument rather than probing `module.scaling` also matters on its own: OPT applies
   `head_size**-0.5` to the query itself and then declares `scaling=1.0`, so probing the attribute
   would scale twice.
2. **The dict index is itself what costs an artifact per layer.** With the instance attached,
   `getattr(module, "attn")` is the same expression in every layer, so one graph can serve the
   whole stack.
3. **The `attention_instances` kwarg no longer has to be drilled down.** With nothing to pass, the
   registry gate relaxes from `is_backend_compatible()` (`_supports_attention_backend`, documented
   upstream as "fully pass the kwargs through all modules up to the Attention layer") to
   `_can_set_attn_implementation()`, which asks only that the model dispatches through the
   interface.

Attaching also makes the backend's module tree the same shape as an in-tree model's, so anything
that walks it -- `process_weights_after_loading`, a state dict, `model.to(device)` -- finds
attention where it expects to. The attribute is `attn`, the name an in-tree model uses
(`LlamaAttention.attn`), which is also what `maybe_remap_kv_scale_name` writes
(`.self_attn.attn.{k,v}_scale`), so a checkpoint's FP8 KV-cache scales land on it with no mapper
entry. `_vllm_mla_attn` is gone; it was never read outside `create_attention_instances`.

## Test Plan

New tests in `tests/models/transformers/test_backend.py`, alongside the device-free ones already
there. They need no accelerator, no weights and no network: the HF stacks are built from
`AutoConfig.for_model(...)` on the `meta` device.

```bash
pytest -q tests/models/transformers/fusers/
pytest -q tests/models/transformers/test_backend.py \
  -k "attention_dispatch or attention_scale or layer_index"
```

- `test_attention_dispatch_is_matched` -- exactly the decoder layers' attention modules match an
  `AttentionFuser`, over `llama`, `gemma4_text` and `deepseek_v3` (MLA).
- `test_attention_layer_index_is_the_modules_own` -- LongCat Flash gives each decoder layer two
  attention sublayers numbered `2i` and `2i + 1`, so `num_hidden_layers` is twice the length of the
  stack and the enclosing layer's position is not the index the KV cache is keyed by. Also pins
  that `validate` is what excludes a module whose config was not patched to dispatch to vLLM, which
  is how a vision tower is left to Transformers.
- `test_attention_scale_is_the_declared_one` -- the scale is the module's, not the default.
  `gemma4_text` (learnable per-dim query weight, so it declares 1.0) and `deepseek_v3` (yarn
  `mscale`) are what make this more than a tautology.
- `test_attention_scale_is_the_argument_not_the_attribute` -- OPT, where the argument (1.0) and the
  attribute (`head_size**-0.5`) disagree.

The existing fuser suites cover the registry change; `test_linear.py`'s QKV and packed-QKV numerics
tests now drive the real `vllm_attention_forward` through the attached `.attn`.

## Test Result

```text
$ pytest -q tests/models/transformers/fusers/
68 passed, 1 skipped in 9.85s

$ pytest -q tests/models/transformers/test_backend.py \
    -k "attention_dispatch or attention_scale or layer_index"
8 passed, 25 deselected in 1.61s
```

The file's other device-free tests -- embedding replacement and multimodal component marking --
pass alongside them, unchanged (3 passed).

**Not yet run, and needed before merge.** Everything requiring an accelerator: the rest of
`tests/models/transformers/`, `tests/models/test_initialization.py -k Transformers`, and model
evals. This changes attention construction for every model on the backend, so `tests/evals/` or
`vllm bench` results belong here before it is reviewed for correctness. Tensor parallel is also
unexercised.

## Notes for reviewers

- **Overlaps with #52156** (attention sinks), which makes `Attention` instances visible to
  `named_modules()` by registering them in an `nn.ModuleList` rather than attaching them to the
  module that dispatches. The two need reconciling; attaching is the one that also removes the
  dict index from the traced graph and gives kv-scale remapping the name it expects.
- **Discovery is source-level on purpose.** `fx_utils` already records the interface call as a leaf
  node, which would be the obvious hook, but `trace` returns a partial graph on failure and a model
  can stop tracing before the dispatch. Reading the forward's source is also what Transformers
  itself does in `_can_set_attn_implementation`, and it is where the `scaling=` argument has to come
  from regardless.
- **Deliberately not keyed off the projection fusers.** A model can dispatch through the interface
  without its projections fusing at all, in which case it would never have been attached.
- **A layer with no identifiable attention module now raises** rather than silently falling back to
  the dict, as does a layer whose modules claim the same `layer_idx`. The latter was already broken:
  two modules sharing an index would have shared one `Attention`.
- AI assistance was used for this change (Claude Code).

### Duplicate check

`gh pr list --state open --search "transformers backend attention in:title,body"` returns no other
PR changing how the backend builds or reaches its attention layers. #52156 is the only related one
and is called out above.

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [X] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [X] The test plan, such as providing test command.
- [X] The test results, such as pasting the results comparison before and after, or e2e results
- [X] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
</details>

### CODE REVIEW

**Review by hmellor — APPROVED** [review]
If the CI passes this should be good

File: vllm/model_executor/models/transformers/fusers/attention.py [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+        """The layer `module` computes attention for, if it declares one."""
+        layer_idx = getattr(module, "layer_idx", None)
+        return layer_idx if isinstance(layer_idx, int) else None
+
+    def scale(self, module: nn.Module) -> float | None:
+        """The softmax scale `module` passes to the interface, or `None`."""
+        if self.scale_expr is None:
+            return None
+        scale = _resolve(self.scale_expr, module)
+        return float(scale) if isinstance(scale, (int, float)) else None
```
[the remark this answers is not in the document]
Comment (bohnstingl) [reply vllm/model_executor/models/transformers/fusers/attention.py]: I integrated this change by raising an Exception if the `scale_expr` exists but cannot be resolved. If necessary, we can handle any default fallbacks in a follow-up PR.

**Comment by hmellor** [comment]
Could we change the supports backend condition to `or` both conditions so maximum permissibility

**Comment by bohnstingl** [comment]
@hmellor done

**Comment by hmellor** [comment]
I've added the recursion back because it's necessary
